segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/ExtS3IntegrationTest.java

```diff
@@ -0,0 +1,359 @@
+        } catch (IOException e) {
+        }
+        result.setObjects(list);
+        return result;
+    }
+
+    @Override
+    public S3ObjectMetadata getObjectMetadata(String bucketName, String key) {
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        EXTS3StorageTest.AclSize data = ACL_MAP.get(key);
+        if (data == null) {
+            throw new S3Exception("NoSuchKey", 0, "NoSuchKey", "");
+        }
+        metadata.setContentLength(data.getSize());
+        Path path = Paths.get(this.baseDir, bucketName, key);
+        metadata.setLastModified(new Date(path.toFile().lastModified()));
+        return metadata;
+    }
+
+    @Override
+    public InputStream readObjectStream(String bucketName, String key, Range range) {
+        byte[] bytes = new byte[Math.toIntExact(range.getLast() + 1 - range.getFirst())];
+        Path path = Paths.get(this.baseDir, bucketName, key);
+        FileInputStream returnStream;
+        try {
+            returnStream = new FileInputStream(path.toFile());
+            returnStream.getChannel().read(ByteBuffer.wrap(bytes), range.getFirst());
+            return new ByteArrayInputStream(bytes);
+        } catch (IOException e) {
+            throw new S3Exception("NoSuchKey", 0, "NoSuchKey", "");
+        }
+    }
+
+    @Override
+    public String initiateMultipartUpload(String bucketName, String key) {
+        multipartUploads.put(key, new HashMap());
+        return Integer.toString(multipartUploads.size());
+    }
+
+    @Override
+    public CopyPartResult copyPart(CopyPartRequest request) {
+        Map<Integer, CopyPartRequest> partMap = multipartUploads.get(request.getKey());
+        if (partMap == null) {
+            throw new S3Exception("NoSuchKey", 0, "NoSuchKey", "");
+        }
+        partMap.put(request.getPartNumber(), request);
+        CopyPartResult result = new CopyPartResult();
+        result.setPartNumber(request.getPartNumber());
+        result.setETag(request.getUploadId());
+        return result;
+    }
+
+    @Override
+    public CompleteMultipartUploadResult completeMultipartUpload(CompleteMultipartUploadRequest request) {
+        Map<Integer, CopyPartRequest> partMap = multipartUploads.get(request.getKey());
+        if (partMap == null) {
+            throw new S3Exception("NoSuchKey", 0, "NoSuchKey", "");
+        }
+        partMap.forEach((index, copyPart) -> {
+            if (copyPart.getKey() != copyPart.getSourceKey()) {
+                Path sourcePath = Paths.get(this.baseDir, copyPart.getBucketName(), copyPart.getSourceKey());
+                Path targetPath = Paths.get(this.baseDir, copyPart.getBucketName(), copyPart.getKey());
+                try (FileChannel sourceChannel = FileChannel.open(sourcePath, StandardOpenOption.READ);
+                     FileChannel targetChannel = FileChannel.open(targetPath, StandardOpenOption.WRITE)) {
+                    targetChannel.transferFrom(sourceChannel, Files.size(targetPath),
```

Comment:
That is just for non blocking channels. But file channels are blocking channels.